extensions/amp-gfycat/0.1/amp-gfycat.js

```diff
@@ -48,34 +74,233 @@ class AmpGfycat extends AMP.BaseElement {
   }
 
   /** @override */
-  layoutCallback() {
-    const gfyid = user().assert(
+  buildCallback() {
+    this.videoid_ = this.getVideoId_();
+
+    // Enable autoplay by default
+    if (!this.element.hasAttribute('noautoplay')) {
+      this.element.setAttribute('autoplay', '');
+    }
+
+    installVideoManagerForDoc(this.element);
+    Services.videoManagerForDoc(this.element).register(this);
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    const placeholder = this.win.document.createElement('amp-img');
+    dev().assert(this.videoid_);
+    const videoid = this.videoid_;
+
+    placeholder.setAttribute('src',
+        'https://thumbs.gfycat.com/' +
+        encodeURIComponent(videoid) + '-poster.jpg');
+    placeholder.setAttribute('layout', 'fill');
+    placeholder.setAttribute('placeholder', '');
+    placeholder.setAttribute('referrerpolicy', 'origin');
+    this.applyFillContent(placeholder);
+
+    return placeholder;
+  }
+
+  /**
+   * @return {string}
+   * @private
+   */
+  getVideoId_() {
+    return user().assert(
         this.element.getAttribute('data-gfyid'),
         'The data-gfyid attribute is required for <amp-gfycat> %s',
         this.element);
-    const noautoplay = this.element.hasAttribute('noautoplay');
+  }
 
-    const iframe = this.element.ownerDocument.createElement('iframe');
-    iframe.setAttribute('frameborder', '0');
+  /** @return {string} */
+  getVideoIframeSrc_() {
+    if (this.videoIframeSrc_) {
+      return this.videoIframeSrc_;
+    }
+
+    const videoid = this.videoid_;
+    let src = 'https://gfycat.com/ifr/' + encodeURIComponent(videoid);
 
-    let src = 'https://gfycat.com/ifr/' + encodeURIComponent(gfyid);
+    const params = getDataParamsFromAttributes(this.element);
+
+    const noautoplay = this.element.hasAttribute('noautoplay');
     if (noautoplay) {
-      src += '?autoplay=0';
+      params['autoplay'] = '0';
```

Comment:
hmm, with this PR, AMP completely manages autoplay, I believe you want to set autoplay=0 always now, no?

Reply:
@aghassemi I fixed video id assert.

About `autoplay=0` - yes, AMP manages video autoplay. But an iframe needs this attribute to be 0 only when iframe doesn't autoplay, it uses it for initialization, to set initial state for controls.

Reply:
Also the second build - integration_tests - keeps failing, I'm not sure why. I'm keeping this branch up to date with the latest master.

Reply:
got it, might be worth documenting this reason as code comment
I restarted integration tests, looks like a flake unrelated to your changes.